### PR TITLE
Juniper private config

### PR DESCRIPTION
**Pull request by markgoddard**

### CODE REVIEW

**Comment by markgoddard** [comment]
@pheller I can't seem to add you as a reviewer, but please feel free to comment.

File: networking_generic_switch/tests/unit/netmiko/test_juniper.py [hunk 1]
```diff
@@ -84,6 +85,15 @@ def test_delete_port(self, mock_exec):
             ['delete interface 3333 unit 0 family ethernet-switching '
              'vlan members'])
 
+    def test_send_config_set(self):
+        connect_mock = mock.MagicMock(netmiko.base_connection.BaseConnection)
+        connect_mock.send_config_set.return_value = 'fake output'
+        result = self.switch.send_config_set(connect_mock, ['spam ham aaaa'])
+        self.assertFalse(connect_mock.enable.called)
+        connect_mock.send_config_set.assert_called_once_with(
+            config_commands=['spam ham aaaa'], exit_config_mode=False)
```
Comment (oneswig) [inline networking_generic_switch/tests/unit/netmiko/test_juniper.py]: cute

File: networking_generic_switch/exceptions.py [hunk 2]
```diff
@@ -35,3 +35,7 @@ class GenericSwitchNetmikoNotSupported(GenericSwitchException):
 
 class GenericSwitchNetmikoConnectError(GenericSwitchException):
     message = _("Netmiko connection error: %(config)s, error: %(error)s")
+
+
+class GenericSwitchNetmikoConfigError(GenericSwitchException):
```
Comment (oneswig) [inline networking_generic_switch/exceptions.py]: Have you annotated code paths where this exception could now be thrown?

File: networking_generic_switch/devices/netmiko_devices/juniper.py [hunk 3]
```diff
@@ -97,19 +97,26 @@ def save_configuration(self, net_connect):
...
         :param net_connect: a netmiko connection object.
         """
-
         # Junos configuration is transactional, and requires an explicit commit
         # of changes in order for them to be applied. Since committing requires
         # an exclusive lock on the configuration database, it can fail if
         # another session has a lock. We use a retry mechanism to work around
         # this.
+
+        class DBLocked(Exception):
```
Comment (oneswig) [inline networking_generic_switch/devices/netmiko_devices/juniper.py]: I've now seen this in the wild, the log message is not terrifically helpful.

Comment (markgoddard) [reply networking_generic_switch/devices/netmiko_devices/juniper.py]: The poor error reporting is due to https://github.com/ktbyers/netmiko/issues/601. I backported my fix for the issue and added it to the neutron server container.

File: networking_generic_switch/tests/unit/netmiko/test_juniper.py [hunk 4]
```diff
@@ -51,7 +51,8 @@ def test_add_network_with_trunk_ports(self, mock_exec):
                 'NetmikoSwitch.send_commands_to_device')
     def test_del_network(self, mock_exec):
         self.switch.del_network(33, '0ae071f55be943e480eae41fefe85b21')
-        mock_exec.assert_called_with(['delete vlans 0ae071f55be943e480eae41fefe85b21'])
+        mock_exec.assert_called_with(
+            ['delete vlans 0ae071f55be943e480eae41fefe85b21'])
```
[the remark this answers is not in the document]
Comment (markgoddard) [reply networking_generic_switch/tests/unit/netmiko/test_juniper.py]: Have to keep pep8 happy.

File: networking_generic_switch/tests/unit/netmiko/test_juniper.py (hunk 1 above)
Comment (markgoddard) [reply networking_generic_switch/tests/unit/netmiko/test_juniper.py]: This is the standard NGS test string.

File: networking_generic_switch/exceptions.py (hunk 2 above)
Comment (markgoddard) [reply networking_generic_switch/exceptions.py]: No I haven't, I'll do that.

**Pull request merged**